lua/metals.lua

```diff
@@ -63,6 +63,63 @@ M.sources_scan = function()
   })
 end
 
+M["metals/quickPick"] = function(_, _, resp)
+  local ids = {}
+  local labels = {}
+  for i, item in pairs(resp.items) do
+    table.insert(ids, item.id)
+    table.insert(labels, i .. ' - ' .. item.label )
+  end
+
+  local choice = util.input_list(labels)
+  if (choice == 0) then
+    print("\nmetals: operation cancelled")
+    return { cancelled = true; }
+  else
+    return { itemId = ids[choice] }
+  end
+end
+
+M['metals/inputBox'] = function(_, _, resp)
+    local name = util.input_box(resp.prompt .. ': ')
+
+    if (name == '') then
+      print("\nmetals: operation cancelled")
+      return { cancelled = true; }
+    else
+      return { value = name; }
+    end
+end
+
+M['metals/executeClientCommand'] = function(_, _, cmd_request)
+  if cmd_request.command == 'metals-goto-location' then
+    lsp.util.jump_to_location(cmd_request.arguments[1])
+  end
+end
+
+--[[
+directory_uri_opt: Path URI for the new file. Defaults to current path. e.g. 'file:///home/...'
+name_opt: Name for the scala file. e.g.: 'MyNewClass'. If nil, it's asked in an input box.
+--]]
+M.new_scala_file = function(directory_uri_opt, name_opt)
+  local args_string_array = {}
+  if directory_uri_opt then
+    table.insert(args_string_array, 1, directory_uri_opt)
+  else
+    table.insert(args_string_array, 1, vim.NIL)
+  end
+  if name_opt then
+    table.insert(args_string_array, 2, name_opt)
+  else
+    table.insert(args_string_array, 2, vim.NIL)
+  end
+
+  execute_command({
+      command = 'metals.new-scala-file';
+      arguments = args_string_array
```

Comment:
I didn't realize this before, but on the Metals side, this is a bit odd to me that it's an array of two optional strings rather than 
```ts
{
  uri?: string,
  filename?: string
}
```